### PR TITLE
add ctc_loss impl for cambricon

### PR DESCRIPTION
**Pull request by jingguo-st**
## Motivation and Context

add ctc_loss impl for cambricon

[note]
not all parameters used in the function impl.

## Checklist
**Before PR**:

- [x] I have read and followed the workflow indicated in the [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: impl/camb/functions/ctc_loss.cpp [hunk 1]
```diff
@@ -0,0 +1,208 @@
+/**
+ * @file
+ * @author DeepLink
+ * @copyright  (c) 2023, DeepLink.
+ */
+
+#include <diopi/functions.h>
```
Comment (yangbofun) [inline impl/camb/functions/ctc_loss.cpp:7]: 把这个diopi/functions.h删掉。

File: impl/camb/functions/ctc_loss.cpp [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+                               workspace,
+                               workspaceSize,
+                               lossTensorDesc.get(),
+                               lossTensor.data(),
+                               backward ? gradTensorDesc.get() : nullptr,
+                               backward ? gradTensor.data() : nullptr));
+    return diopiSuccess;
+}
+
+DIOPI_API diopiError_t diopiCTCLoss(diopiContextHandle_t ctx, diopiTensorHandle_t out, diopiTensorHandle_t negLogLikelihood, diopiTensorHandle_t logAlpha,
```
Comment (yangbofun) [inline impl/camb/functions/ctc_loss.cpp:78]: 把DIOPI_API删掉

File: impl/camb/functions/ctc_loss.cpp [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                                    diopiConstTensorHandle_t logProbs, diopiConstTensorHandle_t targets, diopiConstTensorHandle_t inputLengths,
+                                    diopiConstTensorHandle_t targetLengths, int64_t blank, diopiReduction_t reduction, bool zeroInfinity) {
+    // input and nll, la, out.
+    DiopiTensor outTensor(out);
+    DiopiTensor negLLTensor(negLogLikelihood);
+    DiopiTensor logAlphaTensor(logAlpha);
+    DiopiTensor logProbsTensor(logProbs);
+
+    std::vector<DiopiTensor *> inOutTensorsVecPtr{&outTensor, &negLLTensor, &logAlphaTensor, &logProbsTensor};
```
Comment (yangbofun) [inline impl/camb/functions/ctc_loss.cpp:87]: 类型转化，out和input分开写。

File: impl/camb/functions/ctc_loss.cpp (hunk 2 above)
Comment (jingguo-st) [reply impl/camb/functions/ctc_loss.cpp:78]: fixed

File: impl/camb/functions/ctc_loss.cpp [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    inputLengthsTensor = *labelLengthTensorsVecPtr[1];
+    targetLengthTensor = *labelLengthTensorsVecPtr[2];
+
+    // ctc_loss descriptor
+    CnnlResourceGuard<cnnlCTCLossDescriptor_t, cnnlCreateCTCLossDescriptor, cnnlDestroyCTCLossDescriptor> ctcLossDescObj;
+    cnnlCTCLossDescriptor_t ctcLossDesc = ctcLossDescObj.get();
+
+    int maxInputLength = logProbsTensor.shape()[0];
+
+    int32_t *htargetLength = reinterpret_cast<int32_t *>(malloc(sizeof(int32_t) * targetLengthTensor.numel()));
```
Comment (yangbofun) [inline impl/camb/functions/ctc_loss.cpp:177]: 这个变量用智能指针管理吧。这样以防忘了free。

File: impl/camb/functions/ctc_loss.cpp (hunk 3 above)
Comment (jingguo-st) [reply impl/camb/functions/ctc_loss.cpp]: fixed

File: impl/camb/functions/ctc_loss.cpp [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        default:
+            DIOPI_CHECK(false, "The reduction mode does not supported.");
+    }
+    return diopiSuccess;
+}
+
+static diopiError_t maxTensorElement(diopiContextHandle_t ctx, DiopiTensor targetLengthTensor, int32_t *maxTargetLen) {
+    std::vector<int32_t> targetLengthVector(targetLengthTensor.numel(), 0);
+    // syncStreamInCtx(ctx);
+    auto cnrtRet = cnrtMemcpy(targetLengthVector.data(), targetLengthTensor.data(), sizeof(int32_t) * targetLengthTensor.numel(), cnrtMemcpyDevToHost);
```
Comment (yangbofun) [inline impl/camb/functions/ctc_loss.cpp:41]: 改为调用cnrtMemcpyAsync，同步放到调用后。

File: impl/camb/functions/ctc_loss.cpp [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+    std::vector<DiopiTensor *> labelLengthTensorsVecPtr{&targetTensor, &inputLengthsTensor, &targetLengthTensor};
+    std::set<diopiDtype_t> labelLengthSupportedDtype{diopi_dtype_int32, diopi_dtype_int64};
+    DIOPI_CALL(autoCastTensorType(ctx, labelLengthTensorsVecPtr, labelLengthSupportedDtype));
+
+    DiopiTensor gradInputTensorTemp = gradInputTensor;
+    if (gradInputTensor.dtype() != logProbsTensor.dtype()) {
+        DIOPI_CALL(dataTypeCast(ctx, gradInputTensorTemp, logProbsTensor.dtype()));
+    }
+    syncStreamInCtx(ctx);
```
Comment (yangbofun) [inline impl/camb/functions/ctc_loss.cpp:162]: 删除同步。

Comment (jingguo-st) [reply impl/camb/functions/ctc_loss.cpp:162]: fixed

File: impl/camb/functions/ctc_loss.cpp (hunk 5 above)
Comment (jingguo-st) [reply impl/camb/functions/ctc_loss.cpp:41]: fixed

**Review by yangbofun — APPROVED** [review]

**Pull request merged**